.github/actions/job_init/action.yml

```diff
@@ -0,0 +1,72 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+---
+name: "Initializes the job"
+description: "Does all the necessary steps to set up the job"
+inputs:
+  jdk_version:
+    description: "The JDK version that's supposed to be used."
+    required: true
+  maven_repo_folder:
+    description: "The location of the local Maven repository (not setting this parameter will omit the caching of Maven artifacts)."
+    required: false
+    default: ""
+  source_directory:
+    description: "Specifies the directory from which the code should be moved from (needed for containerized runs; not setting this parameter will omit moving the checkout)."
+    required: false
+    default: ""
+  target_directory:
+    description: "Specifies the directory to which the code should be moved to (needed for containerized runs; not setting this parameter will omit moving the checkout)."
+    required: false
+    default: ""
+runs:
+  using: "composite"
+  steps:
+    - name: "Initializes GHA_PIPELINE_START_TIME with the job's start time"
+      shell: bash
+      run: |
+        job_start_time="$(date --rfc-3339=seconds)"
+        echo "GHA_PIPELINE_START_TIME=${job_start_time}" >> "${GITHUB_ENV}"
```

Comment:
GitHub Actions doesn't provide this environment variable (and I didn't find anything similar) to calculate the actual timeout in `uploading_watchdog.sh`. The problem is as well that this start up doesn't include the Docker image pull and the git checkout which take some time.

One thing we could do here is to do a partial initial checkout that only checks out the `.github` subfolder and do a full checkout in the `job_init` composite action. But I opted against it for now. We could do this as a follow-up if we run into issues due to the not so precise calculation of the job's "real timeout time".

Reply:
This should be good enough I think.

> One thing we could do here is to do a partial initial checkout that only checks out the .github subfolder and do a full checkout in the job_init composite action. But I opted against it for now. We could do this as a follow-up if we run into issues due to the not so precise calculation of the job's "real timeout time".

How would this cover the docker image pulling?

Reply:
It wouldn't. Docker image pulling is happening before anything we can configure kicks in as far as I can see.

Reply:
on another thought. We can calculate the start time outside of the template and pass it in as a parameter. That would include all the steps of the job run, I guess. :bulb:

Reply:
Let's not overthink this. In the end it only matters for cases where the setup takes a long time AND the tests would run into a timeout, which should be quite rare.